diff --git a/databaseSchema.sql b/databaseSchema.sql
index c64d1d5..eed5fd7 100644
--- a/databaseSchema.sql
+++ b/databaseSchema.sql
@@ -1,5 +1,5 @@
 CREATE TABLE Subscriptions (
-    SubscriptionId INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+    Id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
     YoutubeId CHAR(64) NOT NULL UNIQUE,
     Kind CHAR(16) NOT NULL CHECK( Kind IN ('Playlist','Channel') ),
     Title TEXT NOT NULL,
@@ -8,7 +8,7 @@ CREATE TABLE Subscriptions (
 );
 
 CREATE TABLE Videos (
-    VideoId INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, 
+    Id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, 
     YoutubeId CHAR(64) NOT NULL UNIQUE, 
     Title TEXT NOT NULL, 
     DurationSec INT NOT NULL,
@@ -16,20 +16,15 @@ CREATE TABLE Videos (
     UploadDate DATE NOT NULL,
     ThumbnailURL TEXT NOT NULL,
     SubscriptionId INTEGER NOT NULL,
-    Viewed BOOLEAN NOT NULL,
-    ViewDate TEXT,
-    FOREIGN KEY(SubscriptionId) REFERENCES Subscriptions(SubscriptionId)
+    FOREIGN KEY(SubscriptionId) REFERENCES Subscriptions(Id)
 );
 
 CREATE TABLE Animes (
-    AnimeId INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, 
+    Id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, 
     MalId INTEGER NOT NULL, /* Identifier on the MyAnimeList website */
     Title TEXT NOT NULL,
     NbEpisodes INTEGER NOT NULL, 
     Genres TEXT NOT NULL, /* Comma separated list of items */
-    Viewed BOOLEAN NOT NULL,
-    NotInterested BOOLEAN NOT NULL,
-    ViewDate DATE,
     ThumbnailURL TEXT NOT NULL,
     CurrentStatus CHAR(16) NOT NULL CHECK (CurrentStatus IN ('Completed', 'InProgress', 'Planned')),
     Synopsis TEXT NOT NULL
@@ -38,8 +33,8 @@ CREATE TABLE Animes (
 CREATE TABLE RelatedAnimes (
     FirstId INTEGER NOT NULL, 
     SecondId INTEGER NOT NULL,
-    FOREIGN KEY(FirstId) REFERENCES Animes(AnimeId), 
-    FOREIGN KEY(SecondId) REFERENCES Animes(AnimeId)
+    FOREIGN KEY(FirstId) REFERENCES Animes(Id), 
+    FOREIGN KEY(SecondId) REFERENCES Animes(Id)
 );
 
 
@@ -64,6 +59,7 @@ CREATE TABLE VideoStatus (
     VideoId INTEGER NOT NULL, 
     ViewedStatus CHAR(16) NOT NULL CHECK (ViewedStatus IN ('Viewed', 'NotInterested')),
     ViewDate DATE,
+    PRIMARY KEY(UserId, VideoId),
     FOREIGN KEY(UserId) REFERENCES Users(Id), 
     FOREIGN KEY(VideoId) REFERENCES Videos(Id)
 );
@@ -73,6 +69,7 @@ CREATE TABLE AnimeStatus (
     AnimeId INTEGER NOT NULL, 
     ViewedStatus CHAR(16) NOT NULL CHECK (ViewedStatus IN ('Viewed', 'NotInterested')),
     ViewDate DATE,
+    PRIMARY KEY(UserId, AnimeId),
     FOREIGN KEY(UserId) REFERENCES Users(Id), 
     FOREIGN KEY(AnimeId) REFERENCES Animes(Id)
 );
@@ -84,4 +81,14 @@ ALTER TABLE Subscriptions ADD IsFavorite BOOLEAN NOT NULL DEFAULT FALSE;
 CHANGELOG: Adding tables Animes and Related Animes
 
 CHANGELOG: Adding Users, UserSubscriptions, VideoStatus, AnimeStatus tables
+CHANGELOG: Removal of the Viewed and ViewDate columns for Videos and Anime Tables
+INSERT INTO VideoStatus(UserId, VideoId, ViewedStatus, ViewDate) SELECT 0, Id, 'Viewed', ViewDate FROM Videos WHERE Videos.Viewed = TRUE;
+INSERT INTO AnimeStatus(UserId, AnimeId, ViewedStatus, ViewDate) SELECT 0, Id, 'Viewed', ViewDate FROM Animes WHERE Animes.Viewed = TRUE;
+INSERT INTO AnimeStatus(UserId, AnimeId, ViewedStatus, ViewDate) SELECT 0, Id, 'NotInterested', ViewDate FROM Animes WHERE Animes.Viewed = FALSE AND Animes.NotInterested = TRUE;
+ALTER TABLE Videos DROP COLUMN Viewed;
+ALTER TABLE Videos DROP COLUMN ViewDate;
+ALTER TABLE Animes DROP COLUMN Viewed;
+ALTER TABLE Animes DROP COLUMN ViewDate;
+ALTER TABLE Animes DROP COLUMN NotInterested;
+
 **/
\ No newline at end of file
diff --git a/jobs.js b/jobs.js
index b6ca8a9..fa13075 100644
--- a/jobs.js
+++ b/jobs.js
@@ -50,7 +50,7 @@ async function retrieveYoutubeData() {
 schedule.scheduleJob('0 17 * * *', retrieveYoutubeData);
 
 async function retrieveMALData() {
-    const animes = await db.listViewedAnimes();
+    const animes = await db.listAnimes();
     for (var anime of animes) {
         try {
             console.log(`Updating data for ${anime.Title}`);
@@ -58,8 +58,8 @@ async function retrieveMALData() {
             const newStatus = mal.convertAnimeStatus(animeData["status"]);
             await db.updateAnimeStatus(anime.Id, newStatus);
             for (var relatedAnimeData of animeData["related_anime"]) {
-                const animeInDb = await db.malAnimeIsPresent(relatedAnimeData["node"]["id"]);
-                if (!animeInDb) {
+                const animeInfo = await db.getAnimByMalId(relatedAnimeData["node"]["id"]);
+                if (animeInfo === null) {
                     await addAnimeToDatabase(relatedAnimeData["node"]["id"]);
                 }
             }
diff --git a/routes/animes.js b/routes/animes.js
index af0509a..841cf7e 100644
--- a/routes/animes.js
+++ b/routes/animes.js
@@ -8,17 +8,17 @@ const router = express.Router();
 
 router.post("/listViewed", 
     assertAuth,
-    async (req, res) => ok(res, await db.listViewedAnimes())
+    async (req, res) => ok(res, await db.listViewedAnimes(req.session.userId))
 )
 
 router.post("/listUpcoming",
     assertAuth, 
-    async (req, res) => ok(res, await db.listUpcomingAnimes())
+    async (req, res) => ok(res, await db.listUpcomingAnimes(req.session.userId))
 )
 
 router.post("/listSuggested", 
     assertAuth,
-    async (req, res) => ok(res, await db.listSuggestedAnimes())
+    async (req, res) => ok(res, await db.listSuggestedAnimes(req.session.userId))
 )
 
 router.post("/add", 
diff --git a/services/db.js b/services/db.js
index 2479fe5..08baf8f 100644
--- a/services/db.js
+++ b/services/db.js
@@ -7,54 +7,141 @@ const db = new Database('data.db');
 export async function clearDB() {
     db.prepare("DELETE FROM Videos").run(); 
     db.prepare("DELETE FROM Subscriptions").run(); 
+    //TODO other tables to clear
 }
 
 const listSubStatement = db.prepare("SELECT * FROM Subscriptions");
-export async function getAllSubscriptions() {
-    return listSubStatement.all();
+const listSubStatementForUser = db.prepare("SELECT Subscriptions.* FROM Subscriptions INNER JOIN UserSubscriptions ON Subscriptions.Id = UserSubscriptions.ChannelId WHERE UserSubscriptions.UserId = ?");
+export async function getAllSubscriptions(userId = null) {
+    if (userId === null) {
+        return listSubStatement.all();
+    }
+    else {
+        listSubStatementForUser.all(userId)
+    }
+}
+
+const updateIconStatement = db.prepare("UPDATE Subscriptions SET IconURL = ? WHERE YoutubeId = ?");
+export async function updateSubscriptionIcon(youtubeId, iconUrl) {
+    updateIconStatement.run(iconUrl, youtubeId); 
+}
+
+const addVideoStatement = db.prepare("INSERT INTO Videos (YoutubeId, Title, DurationSec, Details, UploadDate, ThumbnailURL, SubscriptionId, Viewed, ViewDate) " +
+    "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?); ")
+export async function addVideo(youtubeId, title, durationSec, details, uploadDate, thumbnailURL, subscriptionId, viewed) {
+    // The viewed * 1 is because better-sqlite3 does not handle booleans
+    const res = addVideoStatement.run(youtubeId, title, durationSec, details, uploadDate.toISOString(), thumbnailURL, 
+        subscriptionId, viewed * 1, null); 
+    return res.lastInsertRowid;
+}
+
+const listAnimesStatement = db.prepare("SELECT * FROM Animes");
+export async function listAnimes() {
+    return listAnimesStatement.all();
+}
+
+const getAnimByMalIdStatement = db.prepare("SELECT * FROM Animes WHERE MalId = ?")
+export async function getAnimByMalId(malId) {
+    return getAnimByMalIdStatement.get(malId) || null;
+}
+
+
+const listViewedAnimesStatement = db.prepare("SELECT Animes.* FROM Animes INNER JOIN AnimeStatus ON Animes.Id = AnimeStatus.AnimeId WHERE UserId = ? AND ViewedStatus = 'Viewed'");
+export async function listViewedAnimes(userId) {
+    return listViewedAnimesStatement.all(userId);
 }
 
+//TODO: this is broken and needs to be fixed.
+const listUpcomingAnimesStatement = db.prepare("SELECT Animes.* FROM Animes INNER JOIN AnimeStatus ON Animes.Id = AnimeStatus.AnimeId WHERE ViewedStatus = NULL AND CurrentStatus != 'Completed'");
+export async function listUpcomingAnimes(userId) {
+    return listUpcomingAnimesStatement.all(userId);
+}
+
+//TODO: this is broken and needs to be fixed.
+const listSuggestedAnimesStatement = db.prepare("SELECT Animes.* FROM Animes INNER JOIN AnimeStatus ON Animes.Id = AnimeStatus.AnimeId WHERE Viewed = FALSE AND CurrentStatus = 'Completed'");
+export async function listSuggestedAnimes(userId) {
+    return listSuggestedAnimesStatement.all(userId);
+}
+
+
+
+/*
+
 const subDetailsStatement = db.prepare("SELECT * FROM Subscriptions WHERE Id = ?");
 export async function getSubscription(channelId) {
-    return subDetailsStatement.get(channelId);
+    const res = subDetailsStatement.get(channelId);
+    if (res === undefined) {
+        return null;
+    }
+    else {
+        return res;
+    }
+}
+
+const subDetailsYoutubeStatement = db.prepare("SELECT * FROM Subscriptions WHERE Id = ?");
+export async function getSubscriptionForYoutubeId(youtubeId) {
+    const res = subDetailsYoutubeStatement.get(youtubeId);
+    if (res === undefined) {
+        return null;
+    }
+    else {
+        return res;
+    }
 }
 
-const addSubStatement = db.prepare("INSERT INTO Subscriptions(YoutubeId, Kind, Title, IconURL) VALUES (?, ?, ?, ?)");
-export async function addSubscription(youtubeId, kind, title, iconURL) {
-    const res = addSubStatement.run(youtubeId, kind, title, iconURL);
+const checkSubExists = db.prepare("SELECT * FROM Subscriptions WHERE YoutubeId = ? LIMIT 1");
+const insertSubscriptionStatement = db.prepare("INSERT INTO Subscriptions(YoutubeId, Kind, Title, IconURL) VALUES (?, ?, ?, ?)");
+async function insertSubscription(youtubeId, kind, title, iconURL) {
+    let resCheck = checkSubExists.all(youtubeId);
+    if (resCheck.length > 0) {
+        return resCheck[0].Id;
+    }
+
+    const res = insertSubscriptionStatement.run(youtubeId, kind, title, iconURL);
     return res.lastInsertRowid;
 }
 
-const updateIconStatement = db.prepare("UPDATE Subscriptions SET IconURL = ? WHERE YoutubeId = ?");
-export async function updateSubscriptionIcon(youtubeId, iconUrl) {
-    updateIconStatement.run(iconUrl, youtubeId); 
+const addSubscriptionStatement = db.prepare("INSERT INTO UserSubscriptions(UserId, ChannelId, Favorite) VALUES (?, ?, FALSE)");
+export async function addSubscription(userId, youtubeId, kind, title, iconURL) {
+    let subId = await insertSubscription(youtubeId, kind, title, iconURL);
+    if (await isSubscribedTo(userId, subId)) {
+        //Nothing to do, user already subscribed
+        return;
+    }
+
+    const res = addSubscriptionStatement.run(userId, subId);
+    return res.lastInsertRowid;
 }
 
-export async function removeSubscription(id) {
-    //Should also remove all videos associated with this subscription
-    //TODO: sql queries
+const removeSubscriptionStatement = db.prepare("DELETE FROM UserSubscriptions WHERE UserId = ? AND ChannelId = ?")
+export async function removeSubscription(userId, subId) {
+    removeSubscriptionStatement.run(userId, subId);
 }
 
-const isSubStatement = db.prepare("SELECT COUNT(*) AS Result FROM Subscriptions WHERE YoutubeId = ?");
-export async function isSubscribedTo(youtubeId) {
-    return isSubStatement.get(youtubeId).Result != 0;
+const containsSubStatement = db.prepare("SELECT COUNT(*) AS Result FROM Subscriptions WHERE YoutubeId = ?");
+export async function containsSub(youtubeId) {
+    return containsSubStatement.get(youtubeId).Result != 0;
+}
+
+const isSubStatement = db.prepare("SELECT COUNT(*) AS Result FROM UserSubscriptions WHERE UserId = ? AND ChannelId = ?");
+export async function isSubscribedTo(userId, subId) {
+    return isSubStatement.get(userId, subId).Result != 0;
 }
 
 const listVideoStatement = db.prepare("SELECT * FROM Videos WHERE SubscriptionId = ? ORDER BY UploadDate DESC");
-const listViewedVideoStatement = db.prepare("SELECT * FROM Videos WHERE SubscriptionId = ? AND Viewed = ? ORDER BY UploadDate DESC");
-export async function listVideosForSubscription(channelId, viewedCondition) {
+const listViewedVideoStatement = db.prepare("SELECT Videos.* FROM Videos INNER JOIN UserSubscriptions ON Videos.SubscriptionId = UserSubscriptions.SubscriptionId WHERE Videos.SubscriptionId = ? AND UserSubscriptions.ViewedStatus = ? ORDER BY Videos.UploadDate DESC");
+export async function listVideosForSubscription(channelId, userId, viewAll) {
     if (viewedCondition === null) {
         return listVideoStatement.all(channelId);
     } else {
-        // The viewed * 1 is because better-sqlite3 does not handle booleans
-        return listViewedVideoStatement.all(channelId, viewedCondition * 1);
+        return listViewedVideoStatement.all(userId, channelId, viewedCondition);
     }
 }
 
-const listRecentVidFavStatement = db.prepare("SELECT Videos.* FROM Videos INNER JOIN Subscriptions ON Videos.SubscriptionId = Subscriptions.Id WHERE Viewed = FALSE AND Subscriptions.IsFavorite = ? ORDER BY Videos.UploadDate DESC LIMIT ?;");
-export async function listRecentVideos(favorites, limit) {
+const listRecentVidFavStatement = db.prepare("SELECT Videos.* FROM Videos INNER JOIN Subscriptions ON Videos.SubscriptionId = Subscriptions.Id WHERE ViewedStatus = 'Viewed' AND Subscriptions.IsFavorite = ? ORDER BY Videos.UploadDate DESC LIMIT ?;");
+export async function listRecentVideos(user, favorites, limit) {
     // The viewed * 1 is because better-sqlite3 does not handle booleans
-    return listRecentVidFavStatement.all(favorites * 1, limit);
+    return listRecentVidFavStatement.all(user, favorites * 1, limit);
 }
 
 
@@ -63,23 +150,14 @@ export async function hasVideo(youtubeId) {
     return containsVideoStatement.get(youtubeId).Result != 0;
 }
 
-const addVideoStatement = db.prepare("INSERT INTO Videos (YoutubeId, Title, DurationSec, Details, UploadDate, ThumbnailURL, SubscriptionId, Viewed, ViewDate) " +
-    "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?); ")
-export async function addVideo(youtubeId, title, durationSec, details, uploadDate, thumbnailURL, subscriptionId, viewed) {
-    // The viewed * 1 is because better-sqlite3 does not handle booleans
-    const res = addVideoStatement.run(youtubeId, title, durationSec, details, uploadDate.toISOString(), thumbnailURL, 
-        subscriptionId, viewed * 1, null); 
-    return res.lastInsertRowid;
-}
-
-const markViewedStatement = db.prepare("UPDATE Videos SET Viewed = ?, ViewDate = ? WHERE Id = ?");
-export async function setViewed(videoId, viewed, viewDate) {
-    markViewedStatement.run(viewed * 1, viewDate != null ? viewDate.toISOString() : null, videoId);
+const markViewedStatement = db.prepare("UPDATE VideoStatus SET ViewedStatus = ?, ViewDate = ? WHERE VideoId = ?");
+export async function setViewedStatus(userId, videoId, viewed, viewDate) {
+    markViewedStatement.run(userId, viewedStatus, viewDate != null ? viewDate.toISOString() : null, videoId);
 }
 
-const markFavStatement = db.prepare("UPDATE Subscriptions SET IsFavorite = ? WHERE Id = ?")
-export async function setChannelFavorite(channelId, favorite) {
-    markFavStatement.run(favorite * 1, channelId);
+const markFavStatement = db.prepare("UPDATE UserSubscriptions SET Favorite = ? WHERE UserId = ? AND ChannelId = ?")
+export async function setChannelFavorite(userId, channelId, favorite) {
+    markFavStatement.run(favorite * 1, userId, channelId);
 }
 
 const addAnimeStatement = db.prepare("INSERT INTO Animes(MalId, Title, NbEpisodes, Genres, Viewed, NotInterested, ViewDate, ThumbnailURL, CurrentStatus, Synopsis) VALUES (?,?,?,?,?,?,?,?,?,?);")
@@ -88,7 +166,7 @@ export async function addAnime(malId, title, nbEpisodes, genres, thumbnailURL, c
     return res.lastInsertRowid;
 }
 
-const markAnimeWatchedStatement = db.prepare("UPDATE Animes SET Viewed = ?, ViewDate = ? WHERE Id = ?");
+const markAnimeWatchedStatement = db.prepare("UPDATE AnimeStatus SET ViewedStatus = ?, ViewDate = ? WHERE Id = ?");
 export async function markAnimeViewed(animeId, viewed, viewDate) {
     markAnimeWatchedStatement.run(viewed * 1, viewDate != null ? viewDate.toISOString() : null, animeId);
 }
@@ -118,11 +196,6 @@ export async function updateAnimeStatus(id, newStatus) {
     updateAnimeStatusStatement.run(newStatus, id);
 }
 
-const malAnimeIsPresentStatement = db.prepare("SELECT COUNT(*) AS NbEntries FROM Animes WHERE MalId = ?")
-export async function malAnimeIsPresent(malId) {
-    return malAnimeIsPresentStatement.get(malId).NbEntries > 0;
-}
-
 const addUserStatement = db.prepare("INSERT INTO Users(Email, PasswordHash) VALUES (?, ?)");
 export async function addUser(email, password) {
     let res = addUserStatement.run(email, await bcrypt.hash(password, config["passwords"]["bcrypt_rounds"]));
@@ -171,4 +244,7 @@ export default {
     addUser,
     userExists, 
     checkUserPassword,
-}
\ No newline at end of file
+}
+    */
+
+export default {}
\ No newline at end of file
